Enable AzureDefaultCredential for authenticating against Azure Storage backends (closes #3998)

marijncv opened the pull request:
## What changes are proposed in this pull request?

It enables the user to configure DefaultAzureCredential to authenticate with Azure Storage. Issue: https://github.com/mlflow/mlflow/issues/3998

## How is this patch tested?

Used `mlflow artifacts download -u <artifact_uri>` with the default azure credentials 

## Release Notes

Users can now use the [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-python/tree/azure-storage-blob_12.7.0/sdk/identity/azure-identity#defaultazurecredential) to connect with AzureBlobArtifactRepository.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users can now use the [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-python/tree/azure-storage-blob_12.7.0/sdk/identity/azure-identity#defaultazurecredential) to connect with AzureBlobArtifactRepository.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

marijncv committed:
enable azblob spn creds

Signed-off-by: Marijn Valk <marijncv@hotmail.com>

pogil commented:
@akshaya-a is there someone from MSFT that could take a first look at this? (cc @dbczumar)

akshaya-a commented:
> @akshaya-a is there someone from MSFT that could take a first look at this? (cc @dbczumar)

@pogil sure! will tag you in after

akshaya-a commented on mlflow/store/artifact/azure_blob_artifact_repo.py:
```diff
@@ -38,10 +43,25 @@ def __init__(self, artifact_uri, client=None):
+        ):
+            from azure.identity import ClientSecretCredential
+
+            account_url = "https://{account}.blob.core.windows.net".format(account=account)
```
The account URL construction here is only valid on the public cloud, but it's at least as correct as the current implementation - we can probably revisit this logic in a followup feature for private cloud support

akshaya-a commented on mlflow/store/artifact/azure_blob_artifact_repo.py:
```diff
@@ -38,10 +43,25 @@ def __init__(self, artifact_uri, client=None):
+            and "AZURE_CLIENT_ID" in os.environ
+            and "AZURE_CLIENT_SECRET" in os.environ
+        ):
+            from azure.identity import ClientSecretCredential
```
this is certainly a welcome extension to the functionality here! If you're already pulling in azure-identity, is there any reason you chose not to use [the DefaultAzureCredential ](https://github.com/Azure/azure-sdk-for-python/tree/azure-storage-blob_12.7.0/sdk/identity/azure-identity#defaultazurecredential)to expand authentication types to include managed identities?

marijncv replied:
Great suggestion! I've updated the PR to use the DefaultAzureCredential use instead of the ClientSecretCredential.

akshaya-a commented on docs/source/tracking.rst:
```diff
@@ -718,11 +718,14 @@ Azure Blob Storage
+``AZURE_TENANT_ID`` & ``AZURE_CLIENT_ID`` & ``AZURE_CLIENT_SECRET`` environment variables
+(preferring a connection string over the access key, and the access key over the client credentials),
+so you must set one of these variables on both your client application and your MLflow tracking server.
+Finally, you must run ``pip install azure-storage-blob`` separately (on both your client and the server)
```
unfortunately this is no longer true as of [this commit](https://github.com/mlflow/mlflow/commit/74be64c8c469aa6cce5cebb92fcafea4bd5ad3a9#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7), so ...

akshaya-a commented on docs/source/tracking.rst:
```diff
@@ -718,11 +718,14 @@ Azure Blob Storage
+so you must set one of these variables on both your client application and your MLflow tracking server.
+Finally, you must run ``pip install azure-storage-blob`` separately (on both your client and the server)
+to access Azure Blob Storage using the connection string or access key,
+or ``pip install azure-storage-blob azure-identity`` to access the storage via client credentials;
```
(continued) ... this might be confusing to users if they didn't have to extend their installation already. Consider a more descriptive error on your usage by

```
try:
    from azure.identity import ....
except ImportError:
    <more descriptive error on package to install>
```

marijncv replied:
Agreed. Added import to a try catch block and updated the documentation.

akshaya-a commented on mlflow/projects/backend/local.py:
```diff
@@ -295,6 +295,9 @@ def _get_azure_blob_artifact_cmd_and_envs(artifact_repo):
+        "AZURE_CLIENT_SECRET": os.environ.get("AZURE_CLIENT_SECRET"),
     }
     envs = dict((k, v) for k, v in envs.items() if v is not None)
     return [], envs
```
Not sure what scenario you're tackling, but you might also be interested in mapping [the profile directory](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/_environment.py#L12) in.

This will allow any other local behaviors to map into the container

akshaya-a commented:
LGTM! Mostly minor comments around usability - feel free to address them in this PR or track them in follow-up issues! try/excepting the import is probably the main issue I'd address in this PR itself

marijncv committed:
added az default credential

Signed-off-by: Marijn Valk <marijncv@hotmail.com>

marijncv commented:
> LGTM! Mostly minor comments around usability - feel free to address them in this PR or track them in follow-up issues! try/excepting the import is probably the main issue I'd address in this PR itself

Thanks for the review @akshaya-a! I've updated the PR to according to the points except for the account_url construction. That one is better handled in another issue I believe.

akshaya-a commented:
@pogil / @dbczumar - looks good to me! take a peek at adding azure-identity into the extras, I think it's ok given the rest of that set but probably want to break down that key into [azure], [gcp] etc instead of accumulating more longer term (@eedeleon FYI)

marijncv committed:
Merge branch 'master' of https://github.com/mlflow/mlflow into feature/azblob-spn-creds
Merge branch

marijncv committed:
Update documentation and packages

Signed-off-by: Marijn Valk <marijncv@hotmail.com>

marijncv commented:
Hey @akshaya-a, @pogil & @dbczumar

I've slightly changed the PR since the review comments from @akshaya-a. Removed the azure-identity from the [extras] key and updated the documentation with the correct usage instructions.

Hope you can take another look! The added flexibility of authentication mechanisms would really help our business in our daily workings with mlflow

Neiraj commented:
> Hey @akshaya-a, @pogil & @dbczumar
> 
> I've slightly changed the PR since the review comments from @akshaya-a. Removed the azure-identity from the [extras] key and updated the documentation with the correct usage instructions.
> 
> Hope you can take another look! The added flexibility of authentication mechanisms would really help our business in our daily workings with mlflow

This is Awsome ! And would go a long way to make MLFLOW adoption easier for organizations working primarily on Azure stack, such as ours. Please release this asap.

JappaB commented:
This would help us a lot as well. Was checking this PR out for a while now and almost forgot about it. But now I see that it isn't merged yet. If this could be included asap, that would be a very helpful to us. Thanks in advance!

GlamTam commented:
Would be of great help to us. Please consider merging for the next release.

dbczumar reviewed (approved):
LGTM! Thanks a bunch @marijncv !

dbczumar committed:
Merge remote-tracking branch 'origin/master' into feature/azblob-spn-creds

Signed-off-by: dbczumar <corey.zumar@databricks.com>
